### PR TITLE
AIP-39: Add logical_date to OpenAPI DAGRun schema

### PR DESCRIPTION
**Pull request by uranusjr**
I tried renaming `DagRun.execution_date` to `schedule_date` and it produced a plethora of errors everywhere, so I’m taking a less drastic attempt on this.

This PR adds a `schedule_date` field to the OpenAPI schema, and allow POST-ing a new DagRun with that field (instead of `execution_date`), but not both of them. Both `schedule_date` and `execution_date` fields are present when GET-ing a DagRun.

Unfortunately this is not currently working; I can’t successfully express the above and the POST tests are currently failing to validate. I have very little experience with OpenAPI specification, so I’m creating this as a draft hoping some can provide some advice.

cc @ephraimbuddy

### CODE REVIEW

File: airflow/api_connexion/endpoints/dag_run_endpoint.py [hunk 1]
```diff
@@ -242,24 +242,34 @@ def post_dag_run(dag_id, session):
     except ValidationError as err:
         raise BadRequest(detail=str(err))
 
+    try:
+        schedule_date = post_body.pop("schedule_date")
+    except KeyError:
+        schedule_date = post_body.pop("execution_date")
+
```
Comment (ephraimbuddy) [inline airflow/api_connexion/endpoints/dag_run_endpoint.py]:
```suggestion
   schedule_date = post_body.get("schedule_date", post_body["execution_date"])
```
This way we preserve the content of post_body. Or what do you think?

Comment (uranusjr) [reply airflow/api_connexion/endpoints/dag_run_endpoint.py]: `DagRun` does not accept `schedule_date` as a keyword argument (yet), so either we need to modify `post_body` or we can’t use `DagRun(..., **post_body)`.

File: airflow/api_connexion/schemas/dag_run_schema.py [hunk 2]
```diff
@@ -57,21 +60,43 @@ class Meta:
...
+        execution_date = data.get("execution_date", _MISSING)
+        if schedule_date is execution_date is _MISSING:  # Both missing.
+            data["schedule_date"] = data["execution_date"] = str(timezone.utcnow())
+        elif schedule_date is _MISSING:  # Only schedule_date missing.
+            data["schedule_date"] = data["execution_date"]
+        elif execution_date is _MISSING:  # Only execution_date missing.
+            data["execution_date"] = data["schedule_date"]
+        elif schedule_date != execution_date:  # Both provided but don't match.
+            raise BadRequest(
+                "schedule_date conflicts with executrion_date",
```
Comment (ephraimbuddy) [inline airflow/api_connexion/schemas/dag_run_schema.py]:
```suggestion
                "schedule_date conflicts with execution_date",
```

File: airflow/api_connexion/openapi/v1.yaml [hunk 3]
```diff
@@ -1875,17 +1875,26 @@ components:
...
             field of an existing object, the request fails with an BAD_REQUEST error.
 
             This together with DAG_ID are a unique key.
+          format: date-time
+        execution_date:
           type: string
+          nullable: true
+          description: |
+            The execution date. This is the same as schedule_date, kept for backwards compatibility.
+            A request with both schedule_date and this field will fail with an BAD_REQUEST error.
```
Comment (ephraimbuddy) [inline airflow/api_connexion/openapi/v1.yaml]: Will it fail when both are supplied or when they have different values?

Comment (uranusjr) [reply airflow/api_connexion/openapi/v1.yaml]: You got me here 😆 I was undecided between the two and ended up doing different things in the two places. I’ll change the API spec to match implementation for now (i.e. only fail if values differ).

**Comment by uranusjr** [comment]
This now conflicts with #16352, let’s finish that one first.

**Comment by uranusjr** [comment]
We should also add a test to ensure a DagRun created from connexion populated data interval correctly.

File: airflow/models/dag.py [hunk 4]
```diff
@@ -2097,19 +2097,25 @@ def create_dagrun(
...
+                run_type = DagRunType.from_run_id(run_id)
+        elif run_type and execution_date:  # Generate run_id from run_type and execution_date.
             if not isinstance(run_type, DagRunType):
                 raise ValueError(f"`run_type` expected to be a DagRunType is {type(run_type)}")
             run_id = DagRun.generate_run_id(run_type, execution_date)
-        elif not run_id:
+        else:
             raise AirflowException(
                 "Creating DagRun needs either `run_id` or both `run_type` and `execution_date`"
             )
```
Comment (uranusjr) [inline airflow/models/dag.py:2112]: This is actually fixing an existing but. Previously, when `run_id`, `run_type`, and `execution_date` are *all* provided, the explicit `run_id` value would be ignored and overwritten by an auto-generated value. This fixes the logic and respect the explicit `run_id` value if it’s provided.

**Comment by uranusjr** [comment]
Tests are failing due to #17728.

**Review by ephraimbuddy — APPROVED** [review]
LGTM

**Review by ashb — APPROVED** [review]

**Pull request merged**